src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_sag.py

```diff
@@ -0,0 +1,751 @@
+                        cond_attn = store_processor.attention_probs
+                        # self-attention-based degrading of latents
+                        degraded_latents = self.sag_masking(
+                            pred_x0, cond_attn, t, self.pred_eps_from_noise(latents, noise_pred, t)
+                        )
+                        # forward and give guidance
+                        degraded_pred = self.unet(degraded_latents, t, encoder_hidden_states=prompt_embeds).sample
+                        noise_pred += sag_scale * (noise_pred - degraded_pred)
+
+                # compute the previous noisy sample x_t -> x_t-1
+                latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
+
+                # call the callback, if provided
+                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
+
+        # 8. Post-processing
+        image = self.decode_latents(latents)
+
+        # 9. Run safety checker
+        image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
+
+        # 10. Convert to PIL
+        if output_type == "pil":
+            image = self.numpy_to_pil(image)
+
+        if not return_dict:
+            return (image, has_nsfw_concept)
+
+        return StableDiffusionPipelineOutput(images=image, nsfw_content_detected=has_nsfw_concept)
+
+    def sag_masking(self, original_latents, attn_map, t, eps):
+        # Same masking process as in SAG paper: https://arxiv.org/pdf/2210.00939.pdf
+        bh, hw1, hw2 = attn_map.shape
+        b, latent_channel, latent_h, latent_w = original_latents.shape
+        h = self.unet.attention_head_dim
+        if isinstance(h, list):
+            h = h[-1]
+        map_size = math.isqrt(hw1)
+
+        # Produce attention mask
+        attn_map = attn_map.reshape(b, h, hw1, hw2)
+        attn_mask = attn_map.mean(1, keepdim=False).sum(1, keepdim=False) > 1.0
+        attn_mask = (
+            attn_mask.reshape(b, map_size, map_size).unsqueeze(1).repeat(1, latent_channel, 1, 1).type(attn_map.dtype)
+        )
+        attn_mask = torch.nn.functional.interpolate(attn_mask, (latent_h, latent_w))
+
+        # Blur according to the self-attention mask
+        transform = T.GaussianBlur(kernel_size=9, sigma=1.0)
+        degraded_latents = transform(original_latents)
+        degraded_latents = degraded_latents * attn_mask + original_latents * (1 - attn_mask)
+
+        # Noise it again to match the noise level
+        degraded_latents = self.scheduler.add_noise(degraded_latents, noise=eps, timesteps=t)
+
+        return degraded_latents
+
+    # Modified from diffusers.schedulers.scheduling_ddim.DDIMScheduler.step
+    def pred_x0_from_eps(self, sample, model_output, timestep):
```

Comment:
Checking my understanding on the functions `pred_x0_from_eps` and `pred_eps_from_noise`:

My understanding is given the output of the unet, we need both the predicted unnoised image and noise in the noised image (x0 and epsilon). 

We do return x0 from many schedulers as `pred_original_sample`. Are there any issues I'm not seeing with using the `step` method on the existing scheduler?

On function name, I think the `_from_{eps,noise}` isn't entirely accurate? I.e. what x0/eps are calculated from is dependent on what the unet is trained to output. Would more apt function names be `pred_x0` and `pred_eps`? (Also aren't eps/noise both referring to the same thing in this case?)

Reply:
Another note: is epsilon always calculated the same way from the network output -- i.e. do all schedulers use the same network reparameterization. If not, should the epsilon calculation be a method call on the scheduler?

cc @patrickvonplaten

Reply:
Thanks a lot for your comments! @williamberman 

Having some issues with some higher-order schedulers (such as `PNDMScheduler`) not returning the x0, I need to implement the reparameterization outside of the schedulers. I also have a trivial problem that the existing schedulers who return x_0 (ddim, ddpm, etc) clip the predicted x_0 to [-1, 1] to be the final output, which is not expected in our method in the sense that we have to return back to x_t with the same predicted epsilon.

On the function name, I agree the `_from_{eps,noise}` isn't accurate. Will remove that.

Reply:
Ah ok if that's an issue, agreed that's outside the scope of adding the pipeline. Could you add a comment in the function noting that not all schedulers return x_0 and that the schedulers clip x_0? That would be helpful for the future :)

Reply:
Sure, I've added the line